### PR TITLE
[timer] Fix failing Timer test in CI

### PR DESCRIPTION
**Pull request by townsend2010**
- Use millisecond resolution
- Adapt and improve tests

### CODE REVIEW

**Comment by townsend2010** [comment]
I ran the `Timer.*` tests in a loop on a Mac 10.14 CI VM and it iterated over 500 times without error before I stopped it.  I think this is stable now :wink:

File: tests/test_timer.cpp [hunk 1]
```diff
@@ -37,116 +37,112 @@ struct Timer : public testing::Test
...
 TEST_F(Timer, times_out)
 {
-    mpu::Timer t{duration_cast<seconds>(1ms), [this]() {
-                     {
-                         std::lock_guard<std::mutex> lk{cv_m};
-                         timedout.store(true);
-                     }
+    mpu::Timer t{default_timeout, [this]()
+                 {
+                     timedout.store(true);
```
Comment (ricab) [inline tests/test_timer.cpp:50]: Why remove the lock?

File: tests/test_timer.cpp [hunk 2]
```diff
@@ -37,116 +37,112 @@ struct Timer : public testing::Test
...
                      cv.notify_all();
                  }};
-    std::unique_lock<std::mutex> lk(cv_m);
 
     t.start();
     ASSERT_FALSE(timedout.load()) << "Should not have timed out yet";
 
-    cv.wait_for(lk, 10ms); // Windows CI needs longer...
+    std::unique_lock<std::mutex> lk(cv_m);
```
Comment (ricab) [inline tests/test_timer.cpp:57]: This is not doing anything now, right?

**Review by ricab — COMMENTED** [review]
Thanks for figuring this one out. I just don't get dropped locks. The mutexes get locked, but no one on the other side looks at them?

File: tests/test_timer.cpp (hunk 1 above)
Comment (townsend2010) [reply tests/test_timer.cpp:50]: Having the lock in place here can potentially produce a false positive result in that a timeout could actually occur and call this callback, but `timedout` won't get set until the lock is acquired and the check in the test itself will pass because this is blocked.  I observed this tracing the code.  I hope this makes sense, but it's quite complicated.  We can follow up on chat if it will help.

File: tests/test_timer.cpp (hunk 2 above)
Comment (townsend2010) [reply tests/test_timer.cpp:57]: The only thing it does is allow the main thread to wait on the condition variable and get notified when the callback completes so as not to make the test run any longer than necessary.

File: tests/test_timer.cpp (hunk 1 above)
Comment (ricab) [reply tests/test_timer.cpp:50]: I had another look. According to [this](https://en.cppreference.com/w/cpp/thread/condition_variable), the lock is needed for this to operate properly:

> Even if the shared variable is atomic, it must be modified under the mutex in order to correctly publish the modification to the waiting thread.

Comment (ricab) [reply tests/test_timer.cpp:50]: That said, I think these tests make a few assumptions that may not hold. One issue is line 60 assuming no spurious wakes. Assuming there is no timeout between 54 and 55 is another. Between that and 58 yet another. Locks would be needed for all of those assumptions to work. Here's how I think the test should look:

```cpp
TEST_F(Timer, times_out)
{
    mpu::Timer t{default_timeout, [this]()
                 {
                     {
                         std::unique_lock<std::mutex> lk(cv_m);
                         timedout.store(true);
                     }
                     cv.notify_all();
                 }};

    std::unique_lock<std::mutex> lk(cv_m);
    t.start();
    ASSERT_FALSE(timedout.load()) << "Should not have timed out yet"; // because this is locked

    ASSERT_TRUE(cv.wait_for(lk, default_wait, [this] { return timedout.load(); }))
        << "Did not time out within reasonable timeframe";
}
```

The main thread needs to synchronize with the timer during the whole time it is running, hence the lock before start (and in the timer). That lock is only released by the condition variable itself. I pass a predicate to the wait operation to deal with spurious wakes and check the return for timeouts.

Comment (townsend2010) [reply tests/test_timer.cpp:50]: Ok, I agree that we need a predicate in the `wait_for` and also agree that we need to lock the mutex in the callback.  What I don't agree with, and found during fixing this, is acquiring the lock before calling `start()` and holding it the whole time before calling `wait_for()`.

The reason being is that if the lock is held this whole time, it's still possible that the timer times out, but it will block in the callback waiting for the mutex before it can set `timedout` to `true`.  However, the main test thread will see that `timedout` is `false` (the `ASSERT_FALSE(timedout.load()) << "Should not have timed out yet";`) which will result in a false positive since it really did timeout, it's just that it couldn't set the `bool` yet.

**Comment by townsend2010** [comment]
Hey @ricab!

FYI, I renamed the locks and the timer test fixture.

**Comment by ricab** [comment]
Thanks @townsend2010!

File: tests/test_timer.cpp [hunk 3]
```diff
@@ -32,164 +33,555 @@ using namespace testing;
...
+    });
+
+    mpu::Timer t{default_timeout, [this]() {
                      {
-                         std::lock_guard<std::mutex> lk{cv_m};
+                         std::lock_guard<std::mutex> test_lock{cv_m};
                          timedout.store(true);
                      }
                      cv.notify_all();
                  }};
```
Comment (ricab) [inline tests/test_timer.cpp:283]: This too we should be able to avoid repeating. Maybe make the `Timer` obj an attribute of `TimerTest`? I know there are a few variations, but a generic form should be enough, I think. Perhaps dropping the `timedout` boolean and replacing with a `timedout_count` to fit `Timer.restarts too`.

**Review by ricab — COMMENTED** [review]
Pheww, I burned a few neurons on this one :sweat_smile: Very ingenious, I think you nailed the mocking approach to test the Timer :+1: 

That said, in this form the tests are very verbose, making them really hard to follow. The problem is already inherently difficult (threads, several locks, mock indirections), but the whole copy-pasting with small variations and  long functions sent my head spinning. I believe a lot of that repetition can be factored out.

I was unsure if/how some un-duplication was possible, so had to try hands-on and pushed a branch with proposals (linked inline).

File: tests/test_timer.cpp [hunk 4]
```diff
@@ -32,164 +33,555 @@ using namespace testing;
...
-    std::unique_lock<std::mutex> lk(cv_m);
 
     t.start();
-    ASSERT_FALSE(timedout.load()) << "Should not have timed out yet";
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp:55]: It won't hurt to keep it in, but since the test is basically in charge of when a timeout will occur and thus when the callback is called, keeping this in seemed to not really test anything at this point.

File: tests/test_timer.cpp [hunk 5]
```diff
@@ -32,164 +33,555 @@ using namespace testing;
...
+    MockTimerSyncFuncs::GuardedMock attr{MockTimerSyncFuncs::inject()};
+    MockTimerSyncFuncs* mock_timer_sync_funcs = attr.first;
 };
 } // namespace
 
-TEST_F(Timer, times_out)
+TEST_F(TestTimer, times_out)
 {
-    mpu::Timer t{duration_cast<seconds>(1ms), [this]() {
+    EXPECT_CALL(*mock_timer_sync_funcs, wait_for(_, _, _)).WillOnce([this](auto&, auto& timer_lock, auto&) {
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp:63]: Nice, thanks!  I'll fix up these to include this.

File: tests/test_timer.cpp [hunk 6]
```diff
@@ -32,164 +33,555 @@ using namespace testing;
...
+            // Wait to be notified by the mock notify_all()
+            cv.wait(test_lock, [this] { return test_notify_called.load(); });
+            test_notify_called.store(false);
+
+            // Re-lock the Timer's internal lock before returning
+            timer_lock.lock();
+
+            return std::cv_status::no_timeout;
+        })
+        .WillOnce([](auto&...) { return std::cv_status::timeout; });
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp:246]: Actually, it probably should call `notify_all()` here.

File: tests/test_timer.cpp [hunk 7]
```diff
@@ -32,164 +33,555 @@ using namespace testing;
...
+        // Unlock the Timer's internal lock
+        timer_lock.unlock();
+
+        // Wait for notification from main test thread
+        cv.wait(test_lock, [this] { return test_notify_called.load(); });
+
+        // Re-lock the Timer's internal lock before returning
+        timer_lock.lock();
+
+        return std::cv_status::timeout;
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp:75]: Thanks!  I'll incorporate your commits into this branch.

File: tests/test_timer.cpp [hunk 8]
```diff
@@ -32,164 +33,555 @@ using namespace testing;
...
+    EXPECT_CALL(*mock_timer_sync_funcs, notify_all(_))
+        .WillOnce(Return()) // Handles the stop() in start()
+        .WillOnce(          // Handles the stop() called in the test
+            [this](auto&) {
+                {
+                    std::unique_lock<std::mutex> test_lock(cv_m);
+                    test_notify_called.store(true);
+                }
+                // Notify the mocked wait_for()
+                cv.notify_all();
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp:131]: Ok, sure, will look into that.

File: tests/test_timer.cpp [hunk 9]
```diff
@@ -32,164 +33,555 @@ using namespace testing;
...
     t.pause();
 
-    std::this_thread::sleep_for(timeout + 5ms);
-    lk.lock();
-    ASSERT_FALSE(timedout.load()) << "Should not have timed out yet";
+    {
+        // Wait for the the timer to be "paused" in the mocked wait()
+        std::unique_lock<std::mutex> test_lock(cv_m);
+        cv.wait(test_lock, [this] { return timer_paused.load(); });
+    }
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp:299]: Will look into that.

File: tests/test_timer.cpp [hunk 10]
```diff
@@ -32,164 +33,555 @@ using namespace testing;
...
     t.start();
-    std::this_thread::sleep_for(500ms);
+
+    {
+        // Wait for the the timer to be "running" in the mocked wait_for()
+        std::unique_lock<std::mutex> test_lock(cv_m);
+        cv.wait(test_lock, [this] { return timer_running.load(); });
+    }
 
     t.start();
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp:484]: Yes, this test gave me headaches as well :wink:  Sure, I'll add that to *try* to make it clearer.

File: tests/test_timer.cpp (hunk 3 above)
Comment (townsend2010) [reply tests/test_timer.cpp:283]: Good ideas, I'll look into this as well.

File: tests/test_timer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ricab) [reply tests/test_timer.cpp:55]: Right. I thought it might have documentation value, help the reader follow where the test is. But up to you.

File: tests/test_timer.cpp [hunk 11]
```diff
@@ -32,164 +33,555 @@ using namespace testing;
...
+        cv.wait(test_lock, [this] { return test_notify_called.load(); });
+
+        // Re-lock the Timer's internal lock before returning
+        timer_lock.lock();
+
+        return std::cv_status::timeout;
+    });
+
+    // notify_all is called twice via the stop() in start() and the stop() in the dtor
+    EXPECT_CALL(*mock_timer_sync_funcs, notify_all(_)).Times(2).WillRepeatedly(Return());
```
[the remark this answers is not in the document]
Comment (ricab) [reply tests/test_timer.cpp:79]: I don't think it's needed. If no action is specified for a particular mock call, it will just do the default. In void-returning functions, that is just no-op.

FWIW, in functions returning another type, the mock will try to default construct an object of that type. I think it throws if there is no default constructor for the return type, so an action must be provided in that case.

File: include/multipass/timer.h [hunk 12]
```diff
@@ -50,14 +52,26 @@ class Timer
...
+
+class TimerSyncFuncs : public Singleton<TimerSyncFuncs>
+{
+public:
+    TimerSyncFuncs(const Singleton<TimerSyncFuncs>::PrivatePass&) noexcept;
+
+    virtual void notify_all(std::condition_variable& cv);
+    virtual void wait(std::condition_variable& cv, std::unique_lock<std::mutex>& lock);
+    virtual std::cv_status wait_for(std::condition_variable& cv, std::unique_lock<std::mutex>& lock,
+                                    const std::chrono::duration<int, std::milli>& rel_time);
```
Comment (ricab) [inline include/multipass/timer.h:73]: One more thing I just remembered while looking at other singleton code: these ought to be `const`.

Comment (townsend2010) [reply include/multipass/timer.h:73]: Oh, you're absolutely right!  Thanks for pointing that out!

File: include/multipass/timer.h [hunk 13]
```diff
@@ -50,14 +52,26 @@ class Timer
...
     const std::function<void()> callback;
     TimerState current_state;
     std::thread t;
     std::condition_variable cv;
     std::mutex cv_m;
 };
 
+#define MP_TIMER_SYNC_FUNCS multipass::utils::TimerSyncFuncs::instance()
+
+class TimerSyncFuncs : public Singleton<TimerSyncFuncs>
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply include/multipass/timer.h:65]: Ok, after having chatted about this, that makes sense.  I think for this, we'll just keep it the singleton for now since I'm so far into it:)

File: include/multipass/timer.h (hunk 12 above)
Comment (townsend2010) [reply include/multipass/timer.h]: Done

File: tests/test_timer.cpp (hunk 3 above)
Comment (townsend2010) [reply tests/test_timer.cpp]: Ok, done.

File: tests/test_timer.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp]: Ok, hopefully this whole test is better now and is a little easier to follow.

File: tests/test_timer.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp]: Done.

File: tests/test_timer.cpp [hunk 14]
```diff
@@ -32,164 +33,555 @@ using namespace testing;
...
+
+        timer_paused.store(true);
+        cv.notify_all();
+
+        // Waits until the Timer dtor calls stop()
+        cv.wait(test_lock, [this] { return test_notify_called.load(); });
+        test_notify_called.store(false);
+
+        // Re-lock the Timer's internal lock before returning
+        timer_lock.lock();
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp]: Yep, I incorporated your branch into this directly.

File: tests/test_timer.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp]: I couldn't use a `ONCALL().WillByDefault()` because two separate actions are needed per test, but I did make a general `EXPECT_CALL().WillOnce().WillRepeatedly()` for this.  The unfortunate side effect is that I can't test the number of times it is called.  Let me know what you think.

File: tests/test_timer.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp]: Yep, incorporated your branch.  Thanks!

File: tests/test_timer.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp]: This went away during the refactoring.

File: tests/test_timer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp:55]: Re-added this assertion.

File: tests/test_timer.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp]: I changed all of these types of calls to include the matchers.

**Comment by townsend2010** [comment]
Hey @ricab,

I finally got all of the refactoring done.  I left some of your review comments unresolved as I want to make sure it's what you are expecting.  I appreciate your feedback and help on this to make it better!

File: tests/test_timer.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ricab) [reply tests/test_timer.cpp]: Yep, definitely. Thanks!

File: tests/test_timer.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (ricab) [reply tests/test_timer.cpp]: I see, thanks. I am not sure if you could just `EXPECT_CALL` again, adding only `Times`, but that would be kind of redundant. There is an `Expectation` handle that can be used to hold on to expectations, but this appears to be useful only for `After` clauses (no interface to ask `Times`). 

You might be able to turn `test_notify_called` into `test_notify_count`, which tests could then verify. The var is used for other purposes so maybe another one would be needed. And something else would be needed for the dtor call. Or you could extract into a helper method receiving the cardinality and call it in tests.

I think it's ok as it is, but up to you.

[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_timer.cpp]: Yeah, I thought of the `EXPECT_CALL` again with only adding `Times` and that overrides the other `EXPECT_CALL`'s and uses the default return value, so long story short, it breaks things :wink: 

I think it's fine the way it is too since it's not *really* important to test how many times it is called.

File: tests/test_timer.cpp [hunk 15]
```diff
@@ -32,165 +33,280 @@ using namespace testing;
...
+
+    void wait_for(const std::atomic_bool& target_state)
+    {
+        std::unique_lock<std::mutex> test_lock(cv_m);
+        cv.wait(test_lock, [&target_state] { return target_state.load(); });
+    }
+
+    void notify_test_wait()
+    {
+        test_notify_called.store(true);
```
Comment (ricab) [inline tests/test_timer.cpp:96]: It would be nice to lock inside this function, [because](https://en.cppreference.com/w/cpp/thread/condition_variable):

> Even if the shared variable is atomic, it must be modified under the mutex in order to correctly publish the modification to the waiting thread. 

Then the caller doesn't have to. I understand you sometimes need the lock to endure, so it can be returned. Here's [what I mean ](https://github.com/canonical/multipass/commit/3f81125a3306af7653e4728684d0f6b709560b29) (in a commit because GH suggestions not very suitable for different bits of code).

**Review by ricab — APPROVED** [review]
This looks mint, great job! 

I have only a little suggestion tweak, but up to you.

bors d+

File: tests/test_timer.cpp (hunk 15 above)
Comment (townsend2010) [reply tests/test_timer.cpp:96]: Right, I started doing that, but then thought about the lock being needed afterwards in some cases.  I obviously didn't think to *return* the lock :wink:  Good idea, and I'll add this since this is really what I wanted to do.  Thanks!

**Comment by townsend2010** [comment]
Ok, let's get this in!

bors=ricab

**Comment by townsend2010** [comment]
bors r=ricab
